datafusion/common/src/types/logical.rs

```diff
@@ -0,0 +1,124 @@
+
+use core::fmt;
+use std::{cmp::Ordering, hash::Hash, sync::Arc};
+
+use super::NativeType;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum TypeSignature<'a> {
+    /// Represents a built-in native type.
+    Native(&'a NativeType),
+    /// Represents an arrow-compatible extension type.
+    /// (https://arrow.apache.org/docs/format/Columnar.html#extension-types)
+    ///
+    /// The `name` should contain the same value as 'ARROW:extension:name'.
+    Extension {
+        name: &'a str,
+        parameters: &'a [TypeParameter<'a>],
+    },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum TypeParameter<'a> {
+    Type(TypeSignature<'a>),
+    Number(i128),
+}
+
+/// A reference counted [`LogicalType`]
+pub type LogicalTypeRef = Arc<dyn LogicalType>;
+
+pub trait LogicalType: Sync + Send {
+    fn native(&self) -> &NativeType;
+    fn signature(&self) -> TypeSignature<'_>;
+}
+
+impl fmt::Debug for dyn LogicalType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("LogicalType")
+            .field(&self.signature())
+            .field(&self.native())
+            .finish()
+    }
+}
+
+impl PartialEq for dyn LogicalType {
+    fn eq(&self, other: &Self) -> bool {
+        self.native().eq(other.native()) && self.signature().eq(&other.signature())
+    }
+}
+
+impl Eq for dyn LogicalType {}
+
+impl PartialOrd for dyn LogicalType {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for dyn LogicalType {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.signature()
+            .cmp(&other.signature())
+            .then(self.native().cmp(other.native()))
+    }
+}
+
+impl Hash for dyn LogicalType {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.signature().hash(state);
+        self.native().hash(state);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    #![allow(dead_code)]
+
+    use super::{LogicalType, TypeParameter, TypeSignature};
+    use crate::types::NativeType;
+
+    struct MagicalType {}
+
+    impl LogicalType for MagicalType {
+        fn native(&self) -> &NativeType {
+            &NativeType::Utf8
+        }
+
+        fn signature(&self) -> TypeSignature<'_> {
+            TypeSignature::Extension {
+                name: "MagicalType",
+                parameters: &[TypeParameter::Type(TypeSignature::Native(
+                    &NativeType::Boolean,
+                ))],
+            }
+        }
+    }
+
+    fn test(logical_type: &dyn LogicalType) {
+        match logical_type.signature() {
+            TypeSignature::Extension {
+                name: "MagicalType",
+                parameters:
+                    [TypeParameter::Type(TypeSignature::Native(NativeType::Boolean))],
+            } => {}
+            TypeSignature::Native(NativeType::Binary) => todo!(),
+            _ => unimplemented!(),
+        };
+    }
```

Comment:
I've added a more structured `TypeSignature`. I've encountered some problems while iterating for a design of this structure since I wanted to balance ease of handling and information contained. I think I've ended up with a good baseline that can be easily pattern matched and limits the copying of information. ptal @findepi @jayzhan211